sovrin_node/persistence/idr_cache.py

```diff
@@ -91,18 +94,28 @@ def close(self):
         self._keyValueStorage.close()
 
     def currentBatchCreated(self, stateRoot):
-        self.unCommitted[stateRoot] = OrderedDict(self.currentBatchOps)
+        self.unCommitted.append((stateRoot, OrderedDict(self.currentBatchOps)))
         self.currentBatchOps = []
 
-    def batchRejected(self, stateRoot=None):
-        if stateRoot:
-            self.unCommitted[stateRoot] = OrderedDict(self.currentBatchOps)
-        else:
-            self.currentBatchOps = []
+    def batchRejected(self):
+        # Batches are always rejected from end of `self.unCommitted`
+        self.currentBatchOps = []
+        self.unCommitted = self.unCommitted[:-1]
 
     def onBatchCommitted(self, stateRoot):
```

Comment:
We assume that revert in reverse order is guaranteed by the replica
Because we do not preserve state-root after a pre-prepare in replica, so if we preserve it somewhere then we can pass that to replica and before reverting check it.